Prep for semver bump and document current blocker

jtgeibel opened the pull request:
This partially reverts the revert (#3051) of #2990. While we can't bump the version yet, it is worth recovering these changes and documenting the remaining work.

r? @Turbo87

Turbo87 commented:
looks like CI is unhappy 😢

jtgeibel commented:
Forgetting to `cargo fmt` strikes again. Should work now.

Turbo87 commented:
you linked https://github.com/rust-lang/crates.io/pull/3094 in the description, but I'm assuming that was a typo, right? 🤔

Turbo87 commented:
I believe https://github.com/steveklabnik/semver/issues/217 might still be blocking this

jtgeibel commented:
@Turbo87 you're right, I meant to link to #3049. And you're right, I think the linked issue is blocking this.

jtgeibel commented:
@Turbo87 I've rolled back the version bump and added a comment pointing to the blocker. I've updated the PR title and description to reflect the new scope.

Turbo87 reviewed (approved):
would it make sense for us to write a failing test for the blocker so that we don't update accidentally?

feel free to r=me :)

jtgeibel commented:
@Turbo87 I've added a test that covers some scenarios where I observed breakage. I'll probably move this test upstream eventually, but it is worth including here for now.

I've also added a command to the `crates-admin` binary, that will test existing versions and dependency requirements in the database against our current version and the latest `semver` release.

jtgeibel commented on src/admin/test_new_semver_release.rs:
```diff
@@ -0,0 +1,71 @@
+        let vers = versions.select(num).load(&conn)?;
+        test_versions(&vers);
+
+        let deps = dependencies.select(req).load(&conn)?;
```
I'm not entirely sure about loading all dependencies in a single query, but this is against the replica database and I expect memory usage to be somewhere around 100 MB so I think this is okay for our current scale.

Turbo87 commented on src/admin/test_new_semver_release.rs:
```diff
@@ -0,0 +1,71 @@
+}
+
+fn test_versions(versions: &[String]) {
+    for version in versions {
+        if let Err(e) = semver::Version::parse(version) {
+            println!("Could not parse `{}` as a semver::Version: {}", version, e);
+        }
+    }
+
```
```suggestion
```

this looks like a duplicate to me, no?

Turbo87 commented on src/admin/test_new_semver_release.rs:
```diff
@@ -0,0 +1,71 @@
+fn test_dependency_predicates(versions: &[String]) {
+    for version in versions {
+        if let Err(e) = semver::VersionReq::parse(version) {
+            println!("Could not parse `{}` as a semver::Version: {}", version, e);
```
```suggestion
            println!("Could not parse `{}` as a semver::VersionReq: {}", version, e);
```

Turbo87 commented on src/admin/test_new_semver_release.rs:
```diff
@@ -0,0 +1,71 @@
+
+        if let Err(e) = semver_next::VersionReq::parse(version) {
+            println!(
+                "Could not parse `{}` as a semver_next::Version: {}",
```
```suggestion
                "Could not parse `{}` as a semver_next::VersionReq: {}",
```

bors commented:
:umbrella: The latest upstream changes (presumably #3391) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).
